keras_nlp/layers/modeling/alibi_bias.py

```diff
@@ -0,0 +1,150 @@
+# Copyright 2023 The KerasNLP Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import math
+
+from keras_nlp.api_export import keras_nlp_export
+from keras_nlp.backend import keras
+from keras_nlp.backend import ops
+
+
+@keras_nlp_export("keras_nlp.layers.AlibiBias")
+class AlibiBias(keras.layers.Layer):
+    """A layer that generates alibi bias
+
+    This layer generates a linear, non-learned bias. Defined and formalized in 
+    [Train Short, Test Long: Attention with Linear Biases Enables Input Length Extrapolation](https://arxiv.org/abs/2108.12409).
+
+    Takes as input an embedded token tensor. The input must have shape
+    `(batch_size, sequence_length, hidden_dim)`. This layer will return an alibi
+    bias of the shape `(1, num_heads, 1, sequence_length)`, which will be added to
+    the result of the query-key dot product in the multi-head attention layer of
+    the transformer. 
+
+    Args:
+        num_heads: int. The number of heads in the multi-head attention layer of
+            the transformer.
+        alibi_bias_max: int. This value will be used to compute the slope of 
+            each head. The heads slopes is a geometric sequence that starts at 
+            `2**(-alibi_bias_max/num_heads)` and uses that same value as its 
+            ratio. Defaults to 8.
+        full: bool. Whether to return the full alibi bias tensor. If set to 
```

Comment:
Right now, our rotary layer applies itself to an input tensor.

```
query = self._rotary_embedding_layer(query)
key = self._rotary_embedding_layer(key)
```

Can we keep a similar usage for abili? Something like
```
attn_scores = self._alibi_layer(attn_scores)
```

Then we would not need to expose `batched` (always broadcast) or `num_heads` (read from input shape) or `full`?

WDYT?

Reply:
Good question re efficiency, and short answer is I don't know. I would expect that overall this is a cheaper part of the forward pass, so practically speed will not be too important. But that is definitely something we should test at some point.

The tensor that we are adding in is fixed right? Nothing is trainable, and recomputing it would be the same each time?

If so, I think we could follow the `RotaryEmbeddng` patter here for now. The `RotaryEmbeddng` faces exactly the same problem, so let's be consistent in our approach. If we discover this is important for actual overall train time, we can update both layers at once.

(I think you already did the update btw, thanks!)